### PR TITLE
Pause workers on insufficient balance error

### LINKED ISSUE #53: Pause the relayer if the operator has no enough funds
When the relayer operator's balance is lower than fees required to send a transaction it makes sense to pause relayer. Then it should automatically continue operations as soon as the balance is topped up. A message that the operator has no enough funds must appear in the logs before the relayer is paused.

### LINKED ISSUE #66: Pool worker stoped processing new jobs after "insufficient funds" error
After "insufficient funds" error new jobs are added to the queue but are not processed:
```
Sep 20 09:35:48 relayer-mpv zk-zkbob-relayer-1/9a43f688f5fb #033[32minfo#033[39m: Updated gasPrice: { maxFeePerGas: '43570000000', maxPriorityFeePerGas: '43570000000' } {"timestamp":"2022-09-20T06:35:48.352Z"}
Sep 20 09:37:36 relayer-mpv zk-zkbob-relayer-1/9a43f688f5fb  #033[34mdebug#033[39m: Fee: 100000000 {"timestamp":"2022-09-20T06:37:36.372Z"}
Sep 20 09:37:36 relayer-mpv zk-zkbob-relayer-1/9a43f688f5fb  #033[34mdebug#033[39m: Native amount: 0 {"timestamp":"2022-09-20T06:37:36.372Z"}
Sep 20 09:37:36 relayer-mpv zk-zkbob-relayer-1/9a43f688f5fb #033[34mdebug#033[39m: TOKENS -2000000000, ENERGY 0, TX DATA {"fee":"5f5e100","nativeAmount":"0","reciever":{"0":191,"1":61,"2":111,"3":131,"4":12,"5":226,"6":99,"7":202,"8":233,"9":135,"10":25,"11":57,"12":130,"13":25,"14":44,"15":217,"16":144,"17":68,"18":43,"19":83}}, MSG VALUE 0 {"timestamp":"2022-09-20T06:37:36.382Z"}
Sep 20 09:37:36 relayer-mpv zk-zkbob-relayer-1/9a43f688f5fb  #033[34mdebug#033[39m: Added job: 320 {"timestamp":"2022-09-20T06:37:36.645Z"}
Sep 20 09:37:48 relayer-mpv zk-zkbob-relayer-1/9a43f688f5fb #033[32minfo#033[39m: Updated gasPrice: { maxFeePerGas: '42050000000', maxPriorityFeePerGas: '42050000000' } {"timestamp":"2022-09-20T06:37:47.879Z"}
```

### PR DESCRIPTION
**Pull request by LyzeOfKiel**
Changes:
* Relayer pauses both queues if `Insufficient balance` error is returned from RPC. After balance is topped-up both queues are resumed. While both workers can pause queues if they see the error, currently processed `sentTxQueue` job (if any) will still be executed infinitely until balance is topped-up. This is required to prevent potential inconsistency, e.g. when transaction was successfully sent previously but we accidentally dropped it during re-send.
Closes #53 
Closes #66 
* Now, only `sendTransaction` method is wrapped in try-catch block in `poolTxWorker` (similar to `sentTxWorker`) as other errors are automatically handled by worker.
* Refactored tests to remove boilerplate code.
* Fixed a typo in `suppressed` word.

### CODE REVIEW

File: zp-relayer/workers/sentTxWorker.ts [hunk 1]
```diff
@@ -203,6 +204,19 @@ export async function createSentTxWorker<T extends EstimationType>(gasPrice: Gas
           await job.update({
             ...job.data,
           })
+        } else if (isInsufficientBalanceError(err)) {
+          // We don't want to take into account last gasPrice increase
+          job.data.prevAttempts.pop()
```
Comment (k1rill-fedoseev) [inline zp-relayer/workers/sentTxWorker.ts:209]: I think it's better to keep the tx hash regardless of error, the same way we treat gas price errors

File: zp-relayer/workers/poolTxWorker.ts [hunk 2]
```diff
@@ -63,64 +64,75 @@ export async function createPoolTxWorker<T extends EstimationType>(
...
-          },
-          {
-            delay: config.sentTxDelay,
-            priority: nonce,
-          }
-        )
-
-        txHashes.push([txHash, sentJob.id as string])
       } catch (e) {
-        logger.error(`${logPrefix} Send TX failed: ${e}`)
```
Comment (k1rill-fedoseev) [inline zp-relayer/workers/poolTxWorker.ts:121]: Should we still log an error, if it's different from insufficient balance one?

File: zp-relayer/utils/helpers.ts [hunk 3]
```diff
@@ -157,3 +159,21 @@ export function withLoop<R>(f: () => Promise<R>, timeout: number, supressedError
     }
   }
 }
+
+export function waitForFunds(web3: Web3, address: string, cb: (balance: BN) => void, minimumBalance: BN) {
+  return promiseRetry(
```
Comment (k1rill-fedoseev) [inline zp-relayer/utils/helpers.ts:164]: Are we using the default interval here? I think it makes sense to retry the request every minute.

File: zp-relayer/workers/sentTxWorker.ts (hunk 1 above)
Comment (LyzeOfKiel) [reply zp-relayer/workers/sentTxWorker.ts:209]: Agree, but I think we should not increase gas price in this case. Maybe just changing last gas price entry in the `prevAttempts` to previous one will be ok?

Comment (k1rill-fedoseev) [reply zp-relayer/workers/sentTxWorker.ts:209]: Sure, should be fine

File: zp-relayer/workers/poolTxWorker.ts (hunk 2 above)
Comment (LyzeOfKiel) [reply zp-relayer/workers/poolTxWorker.ts:121]: In this case error will be logged by `withErrorLog` wrapper
https://github.com/zkBob/zeropool-relayer/blob/b90be9aa5497c52de561ce9e70e57ac3f90c1b5b/zp-relayer/workers/poolTxWorker.ts#L131

File: zp-relayer/utils/helpers.ts (hunk 3 above)
Comment (LyzeOfKiel) [reply zp-relayer/utils/helpers.ts:164]: Updated to 1 minute

File: zp-relayer/utils/helpers.ts [hunk 4]
```diff
@@ -174,6 +174,10 @@ export function waitForFunds(web3: Web3, address: string, cb: (balance: BN) => v
         retry(new Error('Not enough balance'))
       }
     },
-    { forever: true, factor: 1 }
+    {
+      forever: true,
+      factor: 1,
+      maxTimeout: 60000,
```
Comment (k1rill-fedoseev) [inline zp-relayer/utils/helpers.ts:180]: Should we also add `minTimeout: 60000` ?

Comment (LyzeOfKiel) [reply zp-relayer/utils/helpers.ts:180]: Yeah, missed it

**Review by k1rill-fedoseev — APPROVED** [review]

**Comment by LyzeOfKiel** [comment]
I just realised that in this implementation `sentTxWorker` job that runs infinitely until balance top-up will spawn `waitForFunds` promises on each iteration which is not good. I'll try to find a workaround.

**Comment by LyzeOfKiel** [comment]
I think the easiest solution in this case is just to remove `waitForFunds` call from `sentTxWorker` but leaving the error log about insufficient balance. In this case, we will be notified about it and `poolTxWorker` will still stop queues on the next transaction when it will get the error as expected.

**Comment by akolotov** [comment]
@LyzeOfKiel could you please resolve the merge conflict?

**Pull request merged**